### PR TITLE
compiler.compiles: fix silent failure when compiling a built File object

### LINKED ISSUE #11983: compiler.compiles and compiler.links do not accept files generated by configure_file
**Describe the bug**

Trying to use a file generated by `configure_file` with `compiler.compiles` or `compiler.links` results in the following wall:

```
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/mesonbuild/mesonmain.py", line 194, in run
    return options.run_func(options)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/mesonbuild/msetup.py", line 318, in run
    app.generate()
  File "/usr/local/lib/python3.11/site-packages/mesonbuild/msetup.py", line 194, in generate
    self._generate(env)
  File "/usr/local/lib/python3.11/site-packages/mesonbuild/msetup.py", line 239, in _generate
    intr.run()
  File "/usr/local/lib/python3.11/site-packages/mesonbuild/interpreter/interpreter.py", line 2971, in run
    super().run()
  File "/usr/local/lib/python3.11/site-packages/mesonbuild/interpreterbase/interpreterbase.py", line 164, in run
    self.evaluate_codeblock(self.ast, start=1)
  File "/usr/local/lib/python3.11/site-packages/mesonbuild/interpreterbase/interpreterbase.py", line 190, in evaluate_codeblock
    raise e
  File "/usr/local/lib/python3.11/site-packages/mesonbuild/interpreterbase/interpreterbase.py", line 182, in evaluate_codeblock
    self.evaluate_statement(cur)
  File "/usr/local/lib/python3.11/site-packages/mesonbuild/interpreterbase/interpreterbase.py", line 228, in evaluate_statement
    self.evaluate_foreach(cur)
  File "/usr/local/lib/python3.11/site-packages/mesonbuild/interpreterbase/interpreterbase.py", line 472, in evaluate_foreach
    self.evaluate_codeblock(node.block)
  File "/usr/local/lib/python3.11/site-packages/mesonbuild/interpreterbase/interpreterbase.py", line 190, in evaluate_codeblock
    raise e
  File "/usr/local/lib/python3.11/site-packages/mesonbuild/interpreterbase/interpreterbase.py", line 182, in evaluate_codeblock
    self.evaluate_statement(cur)
  File "/usr/local/lib/python3.11/site-packages/mesonbuild/interpreterbase/interpreterbase.py", line 206, in evaluate_statement
    return self.evaluate_if(cur)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/mesonbuild/interpreterbase/interpreterbase.py", line 301, in evaluate_if
    self.evaluate_codeblock(i.block)
  File "/usr/local/lib/python3.11/site-packages/mesonbuild/interpreterbase/interpreterbase.py", line 190, in evaluate_codeblock
    raise e
  File "/usr/local/lib/python3.11/site-packages/mesonbuild/interpreterbase/interpreterbase.py", line 182, in evaluate_codeblock
    self.evaluate_statement(cur)
  File "/usr/local/lib/python3.11/site-packages/mesonbuild/interpreterbase/interpreterbase.py", line 200, in evaluate_statement
    return self.method_call(cur)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/mesonbuild/interpreterbase/interpreterbase.py", line 537, in method_call
    (h_args, h_kwargs) = self.reduce_arguments(node.args)
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/mesonbuild/interpreterbase/interpreterbase.py", line 593, in reduce_arguments
    reduced_pos = [self.evaluate_statement(arg) for arg in args.arguments]
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/mesonbuild/interpreterbase/interpreterbase.py", line 593, in <listcomp>
    reduced_pos = [self.evaluate_statement(arg) for arg in args.arguments]
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/mesonbuild/interpreterbase/interpreterbase.py", line 200, in evaluate_statement
    return self.method_call(cur)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/mesonbuild/interpreterbase/interpreterbase.py", line 550, in method_call
    res = obj.method_call(method_name, args, kwargs)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/mesonbuild/interpreterbase/baseobjects.py", line 94, in method_call
    return method(args, kwargs)
           ^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/mesonbuild/interpreterbase/decorators.py", line 277, in wrapper
    return f(*nargs, **wrapped_kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/mesonbuild/interpreterbase/decorators.py", line 596, in wrapper
    return f(*wrapped_args, **wrapped_kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/mesonbuild/interpreter/compiler.py", line 457, in compiles_method
    result, cached = self.compiler.compiles(code, self.environment,
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/mesonbuild/compilers/compilers.py", line 1287, in compiles
    with self._build_wrapper(code, env, extra_args, dependencies, mode, disable_cache=disable_cache) as p:
  File "/usr/local/Cellar/python@3.11/3.11.4_1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/contextlib.py", line 137, in __enter__
    return next(self.gen)
           ^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/mesonbuild/compilers/compilers.py", line 1279, in _build_wrapper
    with self.cached_compile(code, env.coredata, extra_args=args, mode=mode, temp_dir=env.scratch_dir) as r:
  File "/usr/local/Cellar/python@3.11/3.11.4_1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/contextlib.py", line 137, in __enter__
    return next(self.gen)
           ^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/mesonbuild/compilers/compilers.py", line 887, in cached_compile
    with self.compile(code, extra_args=extra_args, mode=mode, want_output=False, temp_dir=temp_dir) as p:
  File "/usr/local/Cellar/python@3.11/3.11.4_1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/contextlib.py", line 137, in __enter__
    return next(self.gen)
           ^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/mesonbuild/compilers/compilers.py", line 826, in compile
    with open(code.fname, encoding='utf-8') as f:
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
FileNotFoundError: [Errno 2] No such file or directory: 'armv8_external.S'
```

**To Reproduce**

```meson
project('test', 'c')

python3 = import('python').find_installation()

cc = meson.get_compiler('c')

f = configure_file(
      command: [python3, '-c', 'import sys; print(sys.argv[1])', '@0@'.format('prfm   pldl1strm, [x0]')], 
      input: 'meson.build',
      output: 'armv8_external.S',
      capture: true,
)

cc.compiles(
      f,
      args: ['-x', 'assembler'],
      name: 'should succeed under arm64' 
)
```

This should be easy to fix by replacing the following:

https://github.com/mesonbuild/meson/blob/b92858a5beba687478e56850269bc89bc61b3954/mesonbuild/interpreter/compiler.py#L288-L291

with

```py
        if isinstance(code, mesonlib.File):
            self.interpreter.add_build_def_file(code)
            code = mesonlib.File.from_absolute_file(
                code.absolute_path(self.environment.source_dir, self.environment.build_dir))
```

**Expected behavior**

The check should run against the supplied file.

**system parameters**
* Is this a [cross build](https://mesonbuild.com/Cross-compilation.html) or just a plain native build (for the same computer)? **Cross build**
* what operating system (e.g. MacOS Catalina, Windows 10, CentOS 8.0, Ubuntu 18.04, etc.) **macOS Big Sur 11.7.8**
* what Python version are you using e.g. 3.8.0 **3.8.11**
* what `meson --version` **1.1.1**
* what `ninja --version` if it's a Ninja build **1.11.1**

cc @eli-schwartz  for reference

### PR DESCRIPTION
**Pull request by eli-schwartz**
In order to pass a File object down into the compiler impl and compile it, we cannot pass a string with the filename, and we cannot either pass the File object as-is, since it relies on being given Environment attributes to calculate the relative location. So we build a fresh File object as an absolute path.

But the code to do this was totally broken. Instead of using the File method to get an absolute path, we used one that expected to create builddir-relative paths... and then gave it the absolute source dir as the "relative path portion" prefix. This worked by accident as long as it wasn't a built File, but if it was a built file then we intentionally didn't include that prefix -- which was wrong anyway, since we need the build directory!

Use the correct method to get an absolute path in all cases, and emit a warning if it was a built file. This never worked.

Fixes #11983

/cc @amyspark

### CODE REVIEW

**Comment by amyspark** [comment]
Fixes #11983 (filed it a few moments ago)

File: mesonbuild/interpreter/compiler.py [hunk 1]
```diff
@@ -448,9 +448,12 @@ def get_define_method(self, args: T.Tuple[str], kwargs: 'CommonKW') -> str:
     def compiles_method(self, args: T.Tuple['mesonlib.FileOrString'], kwargs: 'CompileKW') -> bool:
         code = args[0]
         if isinstance(code, mesonlib.File):
+            if code.is_built:
+                FeatureNew.single_use('compiler.compiles with built File object', '1.2.0', self.subproject,
```
Comment (dnicolodi) [inline mesonbuild/interpreter/compiler.py:452]: As a user I would be very puzzled by what a "built File object" is. Is there a way we can reword this to be a little easier to grok?

Comment (xclaesse) [reply mesonbuild/interpreter/compiler.py:452]: I agree it's a bit cryptic message. Maybe "with the return value of configure_file()" as I don't think we have other ways to get a built file? Or "with a file created at configure time"?

Also, the message in links_method() is different, shouldn't we have the same message in both cases?

Comment (eli-schwartz) [reply mesonbuild/interpreter/compiler.py:452]: > Also, the message in links_method() adds "It was silently broken and always returned false.", shouldn't we have that here too?

Doesn't it already?

Comment (xclaesse) [reply mesonbuild/interpreter/compiler.py:452]: > Doesn't it already?

Oh yes, misread because this comment thread in between both lines, sorry.

**Comment by xclaesse** [comment]
LGTM
